### PR TITLE
DATA-34/33 - Adding support for tracking logs from S3.

### PR DESCRIPTION
**Pull request by Squirrel18**
## Description:

This PR adds support for processing tracking logs from S3. The pipeline to achieve this is:

- Get and download the tracking log files of the provided date.
- Split the tracking logs by course id into multiple folders.
- Upload each course tracking log file into Google BigQuery using the schema provided by the project.

A new command called: logsfromS3 was added.

## Reviewer:

- [ ] @andrey-canon

### CODE REVIEW

File: edx2bigquery/bqutil.py [hunk 1]
```diff
@@ -775,6 +775,89 @@ def extract_table_to_gs(dataset_id, table_id, gsfn, format=None, do_gzip=False,
...
+def upload_local_data_to_big_query(dataset_id, table_id, schema, course_id, file_name):
+    """
+    Uploads local tracking logs to the provided dataset and table id from
+    the provided course_id tracking log file.
+
+    Args:
+        dataset_id: Valid Google Big Query dataset ID.
+        table_id: Valid Google Big Query table ID of the provided dataset_id.
+        schema: Schema type to use upon the data.
+        project_id: Google Cloud Platform project ID.
```
Comment (andrey-canon) [inline edx2bigquery/bqutil.py]: the method doesn't  receive an argument calls project_id

File: edx2bigquery/bqutil.py [hunk 2]
```diff
@@ -775,6 +775,89 @@ def extract_table_to_gs(dataset_id, table_id, gsfn, format=None, do_gzip=False,
...
+        course_id: Valid course id of the current proccessed course.
+        file_name: Fil name of the archive that has the tracking log data.
```
Comment (andrey-canon) [inline edx2bigquery/bqutil.py]: I suppose it is "File name......

File: edx2bigquery/bqutil.py [hunk 3]
```diff
@@ -775,6 +775,89 @@ def extract_table_to_gs(dataset_id, table_id, gsfn, format=None, do_gzip=False,
...
+        google.cloud.biquery.schema.SchemaField Object.
+    Raises:
+        Exception: When schema argument is empty or None.
+    """
+    if not schema:
+        raise Exception('Data schema was not provided. Unable to upload data to Google BigQuery.')
+
+    for api_schema_field in schema:
+        schema_field = bigquery.schema.SchemaField.from_api_repr(api_schema_field)
+        yield schema_field
```
Comment (andrey-canon) [inline edx2bigquery/bqutil.py]: you can merge  this line with the previous one

File: edx2bigquery/bqutil.py [hunk 4]
```diff
@@ -775,6 +775,89 @@ def extract_table_to_gs(dataset_id, table_id, gsfn, format=None, do_gzip=False,
...
+
+    job_config.schema = list(get_configuration_schema(schema))
+    job_config.write_disposition = 'WRITE_TRUNCATE'
+    job_config.source_format = bigquery.SourceFormat.NEWLINE_DELIMITED_JSON
+    job_config.autodetect = True
+
+    return job_config
+
+
+def get_configuration_schema(schema):
```
Comment (andrey-canon) [inline edx2bigquery/bqutil.py:839]: do you think, this method is necessary ?, you could  verify the schema  before line 831 and generate the list using comprehension list

File: edx2bigquery/bqutil.py [hunk 5]
```diff
@@ -775,6 +775,89 @@ def extract_table_to_gs(dataset_id, table_id, gsfn, format=None, do_gzip=False,
...
+
+    Args:
+        schema: Schema type to use upon the data.
+    Returns:
+        google.cloud.bigquery.job.LoadJobConfig Object.
+    """
+    job_config = bigquery.LoadJobConfig()
+
+    job_config.schema = list(get_configuration_schema(schema))
+    job_config.write_disposition = 'WRITE_TRUNCATE'
```
Comment (andrey-canon) [inline edx2bigquery/bqutil.py]: this should be a global constant

File: edx2bigquery/load_daily_tracking_logs.py [hunk 6]
```diff
@@ -108,5 +113,126 @@ def load_all_daily_logs_for_course(course_id, gsbucket="gs://x-data", verbose=Tr
...
+
+    Returns:
+        Dict containing the tracking log schema.
+    """
+    my_path = os.path.dirname(os.path.realpath(__file__))
+    schema = json.loads(
+        open('{}/schemas/schema_tracking_log.json'.format(my_path)).read()
+    )['tracking_log']
+
+    return schema
```
Comment (andrey-canon) [inline edx2bigquery/load_daily_tracking_logs.py]: the variable shema is not necessary, you can return the value without assign it

File: edx2bigquery/bqutil.py (hunk 4 above)
Comment (Squirrel18) [reply edx2bigquery/bqutil.py:839]: The configuration schema maybe change in the future. So, I will stick to the current implementation.

**Comment by Squirrel18** [comment]
Comments addressed. @andrey-canon

File: edx2bigquery/main.py [hunk 7]
```diff
@@ -504,10 +557,21 @@ def daily_logs(param, args, steps, course_id=None, verbose=True, wait=False):
...
+                    start_date=param.start_date,
+                    end_date=param.end_date,
+                    verbose=verbose,
+                )
+            else:
+                load_daily_tracking_logs.load_all_daily_logs_for_course(
+                    course_id,
+                    use_local_files=param.use_local_files,
+                    gsbucket=edx2bigquery_config.GS_BUCKET,
+                    verbose=verbose, wait=wait,
```
Comment (andrey-canon) [inline edx2bigquery/main.py]: add a line break

File: edx2bigquery/main.py [hunk 8]
```diff
@@ -2118,6 +2199,9 @@ def CommandLine():
         courses = get_course_ids(args)
         run_parallel_or_serial(logs2bq_single, param, courses, args, parallel=args.parallel)
 
+    elif (args.command=='logsfromS3'):
```
Comment (andrey-canon) [inline edx2bigquery/main.py]: why do you use parentheses ?

Comment (Squirrel18) [reply edx2bigquery/main.py]: I don't know, all the commands are like that.

**Review by andrey-canon — APPROVED** [review]

**Pull request merged**